ENH: Define Cifti2 Axes describing the rows/columns of the Cifti2 data

MichielCottaar opened the pull request:
This pull request is designed to add semantic understanding of the Cifti2 file format to nibabel. This is essentially a port from https://github.com/MichielCottaar/cifti as recently suggested to me by @demianw . 

In short, the Cifti2 file format allows for the storage off matrices (or tensors) with a header describing what each row/column means. This meaning can be assigning to every row/column either a voxel or vertex (BrainModel), a group of voxels and/or vertices (Parcel), simply assigning a name (Scalar), a name and a color map (Label), or a starting point and step (Series). All these different options are stored in the same XML structure and hence have the same class representation in the current Cifti2Header file.

This pull request defines a new cifti2_axes.Axis class, which is a superclass of all the options defined above (BrainModel, Parcel, Scalar, Label, and Series). These new class all support:
1. they can be read from Cifti2 header
2. they can be used to write a Cifti2 header
3. they can be easily generated from scratch (either using the __init__ method or using some from_* class methods)
4. they can be easily introspected
5. they support operations like slicing and can be concatenated  through the '+' operator

There is quite a lot going on here. I would welcome any feedback on the current feature set, the API, and whether people think it is a good idea to include this in nibabel at all.

demianw commented:
I think we can also merge this [utils.py](https://github.com/demianw/nibabel/blob/cifti2-tools/nibabel/cifti2/utils.py) from my branch which allows for a better manipulation of indices. What do you think?

MichielCottaar commented:
Thanks, Demian. Your MatrixIndexMap seems quite similar to my Axis class, although rather than using sub-classing to support the various Cifti2 axes types you handle the various types through an if-statement in the __init__. So I would view it as an alternative, simpler implementation for if people think that my implementation is too complex.

demianw commented:
You are right! I didn't look in detail your code yet. I will go over the interfaces more than the implementation and check their simplicity. You code looks very nice at first sight!

effigies commented:
Hi @MichielCottaar @demianw, what's the status of this? I'm inclined to try to release soon, so if this isn't too far away, I can put it in the queue.

demianw commented:
Hey! I meant to do a code review months ago and I was buried in work. I expect to be able to review by nov 1st the latest. Does that sound good?

demianw commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1224 @@
+    Attributes
+    ----------
+    arr : np.ndarray
+        (N, ) typed array with the actual information on each row/column
```
@MichielCottaar  This documentation is a bit unclear. This corresponds more to an abstract class than to a generic object, which, at least to me, makes me think of Generic as in type-parameterised objects.

MichielCottaar replied:
The typed array has been removed and the constructors updated

demianw commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1224 @@
+            (N, ) structured array with for every element a tuple with 3 elements:
+            - vertex index (-1 for voxels)
+            - 3 voxel indices (-1 for vertices)
+            - string (name of brain structure)
```
@MichielCottaar I think that passing the voxel/vertex/structure elements separated as in the attributes presented on line 130 might be an easier to understand interface. The typed arrays with `-1` harcoded for an ignored value seems suboptimal as interface design. @effigies , you have more the hand on what's the interface style in nibabel, what do you think?

Alternatively, this constructor could be hidden and just creating the objects from the factory methods, such as `from_mask`, which have very comfy and clear interfaces, might be a good option.

demianw commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1224 @@
+            - string (name of parcel)
+            - (M, 3) int array with the M voxel indices in the parcel
+            - Dict[String -> (K, ) int array] mapping surface brain structure names to vertex indices
+        affine : np.ndarray
```
same general comment as with the BrainModel class. I thin that @effigies should advise us on wether there's a preferred type of representation for axis elements than the typed array.

MichielCottaar commented:
I just rebased onto master and adjusted the documentation according to your first suggestion.

I agree that the typed array is not the most natural interface. The main reason for doing this was that I was able to move some of the logic to the Axis class. However, in practice I ended up overriding most of these methods anyway. So perhaps I should just use the Axis class as a convenient class for type checking and to define the interface. I could even make it an abstract base class (https://docs.python.org/3.7/library/abc.html) to make this explicit.

demianw commented:
Great reactivity!
I think that the Axis-classes could just hide the constructor and work through the factory functions which seem quite easy to use.  This just needs updating the documentation of the class to indicate the *recommended* way of building them and leave the internal representation hidden.

Agreed that if finally you don't use anything of the Axis class, the ABC scenario is good design.

> I just rebased onto master and adjusted the documentation according to your first suggestion.
> 
> I agree that the typed array is not the most natural interface. The main reason for doing this was that I was able to move some of the logic to the Axis class. However, in practice I ended up overriding most of these methods anyway. So perhaps I should just use the Axis class as a convenient class for type checking and to define the interface. I could even make it an abstract base class (https://docs.python.org/3.7/library/abc.html) to make this explicit.

MichielCottaar commented:
I have removed the underlying typed arrays as I agree that they were mainly confusing. The constructors are now more useful. Next step is to check why the style check in travis is failing.

effigies commented:
@MichielCottaar Might be worth merging master, to make sure it's not something unrelated that we fixed earlier.

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+        BrainModel
+        """
+        nbm = np.sum([bm.index_count for bm in mim.brain_models])
+        voxel = -np.ones((nbm, 3))
```
Given that these are going to be cast to `int`, later:

```suggestion
        voxel = np.full((nbm, 3), fill_value=-1, dtype=int)
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+        """
+        nbm = np.sum([bm.index_count for bm in mim.brain_models])
+        voxel = -np.ones((nbm, 3))
+        vertex = -np.ones(nbm)
```
```suggestion
        vertex = np.full(nbm, fill_value=-1, dtype=int)
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+                if affine is None:
+                    shape = mim.volume.volume_dimensions
+                    affine = mim.volume.transformation_matrix_voxel_indices_ijk_to_xyz.matrix
+                else:
```
You're checking the same `mim.volume.volume_dimensions` and `mim.volume.transformation_matrix_voxel_indices_ijk_to_xyz.matrix` every loop through. I don't think these are going to change on iterations over `mim.brain_models` (unless the API is weirder than I think). You can probably just drop this `else` block altogether.

I'm assuming that the logic here is actually to avoid setting `affine` and `shape` unless there is a volume model in the `mim.brain_models`, and then using common volume information. If it's possible that there's differing volume information, I assume it will be within the `bm` object, not `mim`.

MichielCottaar replied:
I agree that there is no reason for the `else` block. I will get rid of it

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+            mim.append(cifti_bm)
+        return mim
+
+    def iter_structures(self, ):
```
```suggestion
    def iter_structures(self):
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+        return proposed_name
+
+    @property
+    def is_surface(self, ):
```
```suggestion
    def is_surface(self):
```

This pattern appears a lot. Is there a reason for it? Feel free to argue over it, but if you're okay with removing the `, ` from all such instances, I'd prefer it.

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+        if value is not None:
+            value = np.asarray(value)
+            if value.shape != (4, 4):
+                raise ValueError('Affine transformation should be a 4x4 array')
```
There are a lot of validation checks like this that don't get exercised in the tests. Could you add them?

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+                (self.name == other.name).all() and
+                (self.voxel[~self.is_surface] == other.voxel[~other.is_surface]).all() and
+                (self.vertex[~self.is_surface] == other.vertex[~other.is_surface]).all()
+        )
```
This is a tough one to do line-by-line suggestions, so I'll do it old-style:

```Python
        return (
            (self.affine is None or
             np.allclose(self.affine, other.affine) and
             self.volume_shape == other.volume_shape) and
            self.nvertices == other.nvertices and
            np.array_equal(self.name, other.name) and
            np.array_equal(self.voxel[~self.is_surface], other.voxel[~other.is_surface]) and
            np.array_equal(self.vertex[~self.is_surface], other.vertex[~other.is_surface])
        )
```

I dropped `other.affine is None`, because the `xor` case is already covered on L477. If you want to use a specific threshold for comparing affines, you can add the `atol` and `rtol` parameters to `np.allclose`.

Also, is `~obj.is_surface` the correct selector for `obj.vertex` as well as `obj.voxel`?

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+        -------
+        BrainModel
+        """
+        if isinstance(other, BrainModel):
```
For guards like these that cover the whole body, I would suggest inverting and returning the error case, and then de-indenting the rest.

```suggestion
        if not isinstance(other, BrainModel):
            return NotImplemented
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+                        other.volume_shape != shape
+                ):
+                    raise ValueError("Trying to concatenate two BrainModels defined " +
+                                     "in a different brain volume")
```
Can simplify these 10 lines:

```Python
affine, shape = other.affine, other.volume_shape
if self.affine is not None and (
        not np.allclose(affine, self.affine) or
        self.volume_shape != shape):
    raise ValueError(...)
```

MichielCottaar replied:
If the other BrainModel has an affine or volume_shape of None (which is expected if it does not contain any voxels), this change will lead to an error. 

The existing code ensures two things:
1. If only one of the BrainModels has volumetric data (i.e. the affine is not set to None), the local variables affine and volume_shape will be set to that BrainModel.
2. If both BrainModels have volumetric data, they are checked for consistency of the affine and volume_shape.

effigies replied:
Ah, right. Sorry about that.

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+                    raise ValueError("Trying to concatenate two BrainModels with inconsistent " +
+                                     "number of vertices for %s" % name)
+                nvertices[name] = value
+            return type(self)(
```
I think we generally use the `__class__` approach for methods like these:

```suggestion
            return self.__class__(
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+        - structure.BrainStructure object describing the brain structure the element was taken from
+        """
+        is_surface = self.name[index] in self.nvertices.keys()
+        name = 'vertex' if is_surface else 'voxel'
```
Rather than using `getattr` below, I'd suggest using:

```suggestion
        struct = self.vertex if is_surface else self.voxel
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+        """
+        is_surface = self.name[index] in self.nvertices.keys()
+        name = 'vertex' if is_surface else 'voxel'
+        return is_surface, getattr(self, name)[index], self.name[index]
```
And then
```suggestion
        return is_surface, struct[index], self.name[index]
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1330 @@
+            return self.get_element(item)
+        if isinstance(item, string_types):
+            raise IndexError("Can not index an Axis with a string (except for Parcels)")
+        return type(self)(self.name[item], self.voxel[item], self.vertex[item],
```
```suggestion
        return self.__class__(self.name[item], self.voxel[item], self.vertex[item],
```

effigies commented:
Ah. Looks like `numpy.full` didn't exist until numpy 1.8. I'm a little inclined to update the minimum version to 1.8. It's only 2 months younger than 1.7.1. I'll shoot you a quick PR.

MichielCottaar commented:
I could also go back to -np.ones if you prefer. The arrays in this case are only 1-dimensional, so there should be no memory problems (unless people have more than tens of millions of voxels/vertices in a BrainModel).

effigies commented:
You can if you want, but we've been sitting on 1.7.1 for 2 years. I think it's a slow enough walk up the version chain. I'm inclined to go ahead and increase the version, and if we somehow get somebody who can't upgrade past 1.7.1 complaining, the next patch release can revert and replace `full` with a compatibility function like:

```Python
def np_full(shape, fill_value, dtype=None, order='C'):
    if dtype is None:
        dtype = np.array(fill_value).dtype
    ret = np.empty(shape, dtype=dtype, order=order)
    ret.fill(fill_value)
    return ret
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1340 @@
+                             "('second', 'hertz', 'meter', or 'radian'")
+        self._unit = value.upper()
+
+    def extend(self, other_axis):
```
The expectation with a method named `extend` is in-place extension, as in lists. If possible, I'd prefer not to violate these expectations. Could this implementation simply be renamed to `__add__`?

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1340 @@
+                       if item.stop is None else
+                       (item.stop if item.stop >= 0 else self.size + item.stop))
+            if idx_start > self.size:
+                idx_start = self.size - 1
```
In lists:

```Python
>>> a = list(range(5))
>>> a[6:]
[]
```

By the current logic, this should be equivalent to `a[4:]`, or `[4]`. So I think if we want consistency with lists:

```suggestion
                idx_start = self.size
```

MichielCottaar replied:
These lines are meant to deal with the following situation:
>>> a = list(range(5))
>>> a[6::-1]
[4, 3, 2, 1, 0]

In that case the `idx_start` is `self.size - 1`. I will alter the if-statement to by `if idx_start > self.size and step < 0` to clarify this.

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1340 @@
+        """
+        if index < 0:
+            index = self.size + index
+        if index >= self.size:
```
```suggestion
        if index >= self.size or index < 0:
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1340 @@
+            index = self.size + index
+        if index >= self.size:
+            raise IndexError("index %i is out of range for Series with size %i" %
+                             (index, self.size))
```
If you're going to report the index, I'd suggest reporting the originally passed index, so that negative indices don't get transformed before being reported. Otherwise if you have a series with 9 elements and ask for -1, you'll get "index -2 is out of range for Series with size 9".

MichielCottaar replied:
Done

effigies commented:
Actually, if you just want to look through the [CodeCov report](https://codecov.io/gh/nipy/nibabel/pull/641/diff), that highlights what behavior is still untested.

demianw commented:
Hey @MichielCottaar  @effigies! This is great work. I was swamped this past two weeks but I can review this on Wednesday (or maybe sat morning)

effigies commented:
Thanks @demianw. If you do have time on Saturday, I'd appreciate it. I'm hoping to release on Monday, and this would be a nice addition. No worries if you can't. I'm okay with pushing off a few days.

MichielCottaar commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1340 @@
+                if affine is None:
+                    affine = bm.affine
+                    volume_shape = bm.volume_shape
+                else:
```
Fixed

MichielCottaar commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1340 @@
+        """
+        self.name = np.asanyarray(name, dtype='U')
+        if isinstance(label, dict):
+            label = [label] * self.name.size
```
Fixed and added a test for it as well

MichielCottaar commented:
I added @effigies comments, merged with master, and increased the test coverage (mainly by testing the __eq__ methods in much more detail). The tests will still fail as the numpy version is still 1.7.1 instead of 1.8...

effigies commented:
@MichielCottaar This looks great. Thanks so much for all your work, and I'll wait on @demianw's final review.

Apologies for the nit-pick, but could I ask you to format your docstrings according to [numpy conventions](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)? The main thing I see is things like `Tuple[cifti2_axes.Axis]`, which would translate to something like

```Rst
tuple of :class:`.cifti2_axes.Axis`
```

I know this has been a lot of back-and-forth, so if you like, we can merge it as is and I can do a big pass over the docs before the release.

satra commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1322 @@
+                             "3-dimensional (for volumes), not %i-dimensional" % mask.ndim)
+
+    @classmethod
+    def from_surface(cls, vertices, nvertex, name='Cortex'):
```
if it's left as a default, perhaps we use `Unknown`. i don't think the library should make this choice.

MichielCottaar replied:
I've set it to Other (which is the closest equivalant to Unknown allowed by the CIFTI spec)

effigies commented:
@satra I agree that the documentation sorely needs updating. The focus for a bit has been more getting support for the file formats in, and little thought has been given for supporting users outside the downstream projects that often contributed the code. It would be good to try to make a concerted push to update the documentation before the 3.0 release at latest; I'd be interested to hear any strategies people have for dividing the work and organizing the work.

@MichielCottaar If you have the spare cycles to write some documentation in the short term (possibly just porting over https://github.com/MichielCottaar/cifti/blob/master/README.md into our docs structure?), I would be happy to push off release for a week (or two at most). In my experience, the post-release period has a lull where a lot of good intentions can fade, so if we can harness some momentum now, it's worth the delay.

MichielCottaar commented:
I should have time to port the existing README file and update the language/import statements. Would this involve creating a new rst file in doc/source explaining the module and add it to manual.rst (so that is shows up in https://nipy.org/nibabel/manual.html#manual)? Or would it make more sense to add it to the cifti2/__init__.py file (so that it shows up on the https://nipy.org/nibabel/reference/nibabel.cifti2.html#module-nibabel.cifti2)?

Also, appveyor seems to throw some errors in the Minc2 codebase (potentially related to the update to numpy 1.8). Could somebody look into that?

demianw commented:
Great! I’ll review this tomorrow

demianw commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1430 @@
+        return self.name[index], self.meta[index]
+
+
+class Label(Axis):
```
`LabelAxis` might be a slightly more clear name. `Label` might be confusing even if it's within a `cifti2_axis` module. What do you think @effigies ?

MichielCottaar replied:
I would either add Axis to all of the names (`Scalar` could be unclear as well) or to none for consistency, however as we are always within the `cifti2` environment I think it should be clear enough. @effigies, more thoughts on this would be welcome.

demianw replied:
I agree that it’s all or none. I think that often the module naming gets a bit lost  in python’s flexible import system. In nibabel we have naming like `Nifti2Image` when it could just be `Image` within a module `nitfi2` so this issue is considered elsewhere. I agree that the last call should come from @effigies

effigies replied:
Setting aside the question of interpretability when imported as a bare name, which would you find a better description of the object in question: `cifti2.Label` or `cifti2.LabelAxis` (and so on for each `Axis` subclass)? We should consider (our estimate of) the intuition of someone roughly familiar with the CIFTI-2 format but not our data structures. Note we also have `cifti2.Cifti2Label` objects for the XML nodes.

From my perspective as one who's vaguely acquainted with CIFTI-2, and more via code reviews than actual manipulation of files, I'm a little inclined toward `(BrainModel|Parcels|Scalar|Label|Series)Axis`, but I don't feel very strongly about it.

You asked for a last call, so I'd say let's go with @demianw, but if you feel strongly the other way, @MichielCottaar, please feel free to argue further.

I'm going to go through hopefully a last round of review now.

demianw commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1430 @@
+                        check_name, getattr(self, check_name).shape))
+
+    @classmethod
+    def from_mapping(cls, mim):
```
I'd rather use `index_mapping` instead of `mapping`. This is because `mapping` makes me think of the `Mapping` collection and I try to use it with a `dict` object or similar.

MichielCottaar replied:
done

MichielCottaar commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1430 @@
+    -------
+    cifti2.Cifti2Header
+    """
+    axes = list(axes)
```
I index it in a bit (axes[:dim] & axes.index(ax)) and the tee object does not work for that. I did make it a tuple as it does not need to be altered, which should be slightly faster

MichielCottaar commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1430 @@
+        Returns
+        -------
+        tuple with 3 elements
+        - boolean, which is True if it is a surface element
```
Returning one of the 'CIFTI_MODEL_TYPE' makes sense to me. I think distinguishing these two strings should be easy and declaritive enough (e.g., if 'SURFACE' in model_type) without adding an additional function or method.

MichielCottaar commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1430 @@
+        return proposed_name
+
+    @property
+    def is_surface(self):
```
If we are going to rename these, I would go with `surface_mask` and `voxel_mask` (or `volume_mask`), but I'm not really convinced that we need both as they are just the negation of each other.

demianw replied:
I think the cost of having both is negligible and it will enable more declarative programming

effigies commented on nibabel/cifti2/__init__.py:
```diff
@@ -6,14 +6,15 @@
 #
 ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
-"""CIfTI format IO
+"""CIfTI2 format IO
```
Silly pedantic thing, but CIFTI (like GIFTI) doesn't have a lowercase `f`. Would you mind doing a find-and-replace for this? (The `Cifti2` that refers to the prefix of, e.g. `Cifti2Image` or `Cifti2BrainModel` doesn't need updating. Just the format itself.)

```suggestion
"""CIFTI-2 format IO
```

effigies commented on nibabel/cifti2/cifti2.py:
```diff
@@ -1268,6 +1265,40 @@ def get_index_map(self, index):
+
+        Returns
+        -------
+        axis : cifti2_axes.Axis
```
```suggestion
        axis : :class:`.cifti2_axes.Axis`
```

effigies commented on nibabel/cifti2/cifti2.py:
```diff
@@ -1268,6 +1265,40 @@ def get_index_map(self, index):
+
+        Parameters
+        ----------
+        axes : Tuple[cifti2_axes.Axis]
```
```suggestion
        axes : tuple of :class`.cifti2_axes.Axis`
```

effigies commented on nibabel/cifti2/cifti2.py:
```diff
@@ -1297,15 +1328,19 @@ def __init__(self,
             returns an array from ``np.asanyarray``.  It should have a
             ``shape`` attribute or property.
-        header : Cifti2Header instance
+        header : Cifti2Header instance or Sequence[cifti2_axes.Axis]
```
```suggestion
        header : Cifti2Header instance or sequence of :class:`cifti2_axes.Axis`
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1441 @@
+
+>>> from nibabel import cifti2
+>>> bm_cortex = cifti2.BrainModel.from_mask(cortex_mask,
+...                                         brain_structure='cortex_left') # doctest: +SKIP
```
Are you able to generate these (or dummy) masks easily? Since we run doctests on docs like this, it would help to ensure that the docs are providing usable suggestions. If not, we can add this to the list of things to look into when we do a big sweep through the docs.

MichielCottaar replied:
I created some small, fake masks. I guess a better example would load some atlas and select a single ROI to include in the BrainModelAxis, but I am not aware of one easily availble

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1441 @@
+
+    Parameters
+    ----------
+    mim : cifti2.Cifti2MatrixIndicesMap
```
```suggestion
    mim : :class:`.cifti2.Cifti2MatrixIndicesMap`
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1441 @@
+
+    Returns
+    -------
+    subtype of Axis
```
```suggestion
    subclass of :class:`Axis`
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1441 @@
+
+    Returns
+    -------
+    cifti2.Cifti2Header
```
```suggestion
    :class:`.cifti2.Cifti2Header`
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1441 @@
+                 self.volume_shape == other.volume_shape) and
+                self.nvertices == other.nvertices and
+                np.array_equal(self.name, other.name) and
+                np.array_equal(self.voxel[~self.surface_mask], other.voxel[~other.surface_mask]) and
```
Now I'm looking at this, shouldn't the rest of the non-voxel entries be equal, as well:
```suggestion
                np.array_equal(self.voxel, other.voxel) and
```

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1441 @@
+                self.nvertices == other.nvertices and
+                np.array_equal(self.name, other.name) and
+                np.array_equal(self.voxel[~self.surface_mask], other.voxel[~other.surface_mask]) and
+                np.array_equal(self.vertex[self.surface_mask], other.vertex[other.surface_mask])
```
And here:

```suggestion
                np.array_equal(self.vertex, other.vertex)
```

MichielCottaar replied:
I'm not sure, for greyordinates on the surface the voxel values are meaningful. So a BrainModel where they have been set to -1 (my default) or some other value are referreing to the same greyordinates. And similarly for volumetric greyordinates are the same if they have the same voxel indices, irrespetive of their vertex indices.

effigies replied:
Okay.

effigies commented on nibabel/cifti2/cifti2_axes.py:
```diff
@@ -0,0 +1,1441 @@
+        for idx_parcel, (parcel_name, bm) in enumerate(named_brain_models):
+            all_names.append(parcel_name)
+
+            voxels = bm.voxel[~bm.surface_mask]
```
Might be more intuitive to switch from `~x.surface_mask` to `x.volume_mask`, now that it exists:

```suggestion
            voxels = bm.voxel[bm.volume_mask]
```

MichielCottaar commented:
Any ideas about the appveyor errors, @effigies ? They only happened in python 3.4 for the 32-bit system and seem unrelated with the changes I made in the code.

effigies commented:
@MichielCottaar Those errors are stochastic but persistent. I'm currently not worrying about them. Thanks for this.

@demianw Any final comments?

demianw commented:
Nope! Great job @michielcottaar!

effigies commented:
Thanks for your patience.

MichielCottaar commented:
Thanks for all your great feedback.

demianw commented:
Yay!!!!!

satra commented:
awesome job @MichielCottaar .  this will help us move things forward. i asked @effigies if we can start doing some interactive tutorials for nibabel that would serve as both static tutorials but also something people can immediately start playing with online. we can use the nipype tutorials as an example: https://miykael.github.io/nipype_tutorial/
